WIP: Validator Power Dec-> Int

rigelrozanski opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->
closes #2513

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added entries in `PENDING.md` with issue # 
- [ ] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

rigelrozanski commented on x/stake/keeper/slash.go:
```diff
@@ -29,7 +29,9 @@ func (k Keeper) Slash(ctx sdk.Context, consAddr sdk.ConsAddress, infractionHeigh
 	// Amount of slashing = slash slashFactor * power at time of infraction
-	slashAmount := sdk.NewDec(power).Mul(slashFactor)
+	slashAmountDec := sdk.NewDec(power).Mul(slashFactor)
+	slashAmount := slashAmountDec.TruncateInt()
```
Note this is a truncate

rigelrozanski commented on x/stake/keeper/slash.go:
```diff
@@ -143,29 +145,30 @@ func (k Keeper) Unjail(ctx sdk.Context, consAddr sdk.ConsAddress) {
 	// Calculate slash amount proportional to stake contributing to infraction
-	slashAmount = slashFactor.MulInt(unbondingDelegation.InitialBalance.Amount)
+	slashAmountDec := slashFactor.MulInt(unbondingDelegation.InitialBalance.Amount)
+	slashAmount = slashAmountDec.TruncateInt()
```
Note this is a truncate

rigelrozanski commented on x/stake/keeper/slash.go:
```diff
@@ -189,29 +192,30 @@ func (k Keeper) slashUnbondingDelegation(ctx sdk.Context, unbondingDelegation ty
 	// Calculate slash amount proportional to stake contributing to infraction
-	slashAmount = slashFactor.MulInt(redelegation.InitialBalance.Amount)
+	slashAmountDec := slashFactor.MulInt(redelegation.InitialBalance.Amount)
+	slashAmount = slashAmountDec.TruncateInt()
```
Note this is a Truncate

rigelrozanski commented on x/stake/types/validator.go:
```diff
@@ -386,25 +386,30 @@ func (v Validator) AddTokensFromDel(pool Pool, amount sdk.Int) (Validator, Pool,
+
+	// leave excess tokens in the validator
+	// however burn all the delegator shares
+	issuedTokens := issuedTokensDec.TruncateInt()
```
note these fundamental changes

cwgoes replied:
at first glance this seems fine, we need to be careful about interactions with fee distribution but I don't think truncating and leaving in the validator vs. truncating and moving to the community pool (previous behaviour) changes the attack surface

rigelrozanski replied:
Yeah I don't think this is a problem, just wanted to highlight this change

rigelrozanski commented on x/stake/simulation/invariants.go:
```diff
@@ -94,17 +92,22 @@ func SupplyInvariants(ck bank.Keeper, k stake.Keeper,
 		)
 
+		// XXX TODO investigate why go test TestGaiaSimulationAfterImport
+		//          loses decimal fractions if RoundInt is not used below here
```
Note this inconsistency

rigelrozanski replied:
(error is in the order of the final decimal places)

cwgoes replied:
Might be fixed by changes in https://github.com/cosmos/cosmos-sdk/pull/2982

cwgoes replied:
I think we should not round... safer to merge with simulation_after_import failing but the rest passing

rigelrozanski replied:
k I just undid the rounding

rigelrozanski replied:
woot! this seems to have resolved the issue with no more rounding 😎

cwgoes commented on types/int.go:
```diff
@@ -386,6 +408,16 @@ func (i Uint) IsZero() bool {
 	return i.i.Sign() == 0
 }
 
+// IsNegative returns true if Uint is negative
```
Huh when would a `Uint` be negative? Sounds like a panic should happen...

rigelrozanski replied:
lol - good point wow. just deleting for now, I just wanted to have completion, but obviously this should never be used to begin with

rigelrozanski commented:
@cwgoes - thanks for explaining #2513, yeah this doesn't resolve that issue, but I reckon will make solving this issue a bit smoother. I've changed the description header to REF for #2513, and added closes to #2327 (which was the intended issue)

cwgoes commented:
All the tests seem to fail, is there a known issue or an unknown issue?

alexanderbez commented on types/int.go:
```diff
@@ -386,6 +408,11 @@ func (i Uint) IsZero() bool {
 }
 
+// IsPositive returns true if Uint is positive
+func (i Uint) IsPositive() bool {
```
When would we actually ever call this? I imagine most cases will be `if !foo.IsZero()` since it's an unsigned int.

rigelrozanski replied:
unnecessary removing.

rigelrozanski commented:
@cwgoes I made a small oopsie - all fixed

rigelrozanski commented on x/stake/types/validator.go:
```diff
@@ -386,28 +386,40 @@ func (v Validator) AddTokensFromDel(pool Pool, amount sdk.Int) (Validator, Pool,
+		issuedTokens = v.DelegatorShareExRate().Mul(delShares).TruncateInt()
+	}
+
+	v.Tokens = v.Tokens.Sub(issuedTokens)
```
sure - I'll can move into the if-else statement just to keep that obvious

cwgoes commented:
Also needs a rebase since the changes for the gaia-9002 fix PR.

cwgoes commented:
@rigelrozanski If you have a chance to rebase this I think we can merge it.

rigelrozanski commented:
ready to be merged! so long as no CI fails unexpectedly here

cwgoes commented on x/stake/types/validator.go:
```diff
@@ -383,29 +383,48 @@ func (v Validator) AddTokensFromDel(pool Pool, amount sdk.Int) (Validator, Pool,
+		// leave excess tokens in the validator
+		// however fully use all the delegator shares
+		issuedTokens = v.DelegatorShareExRate().Mul(delShares).TruncateInt()
+		issuedTokens = sdk.MinInt(issuedTokens, v.Tokens) // defensive should not occur
```
If it shouldn't occur then shouldn't we panic?

rigelrozanski replied:
Yup I agree - will make a panic
